Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift

```diff
@@ -0,0 +1,358 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+//
+//  HTTPServerPipelineHandlerTest.swift
+//  NIOHTTP1Tests
+//
+//  Created by Cory Benfield on 02/03/2018.
+//
+
+import XCTest
+import NIO
+import NIOHTTP1
+
+private final class ReadRecorder: ChannelInboundHandler {
+    typealias InboundIn = HTTPServerRequestPart
+
+    enum Event: Equatable {
+        case read(InboundIn)
+        case halfClose
+
+        static func ==(lhs: Event, rhs: Event) -> Bool {
+            switch (lhs, rhs) {
+            case (.read(let b1), .read(let b2)):
+                return b1 == b2
+            case (.halfClose, .halfClose):
+                return true
+            default:
+                return false
+            }
+        }
+    }
+
+    public var reads: [Event] = []
+
+    func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        self.reads.append(.read(self.unwrapInboundIn(data)))
+    }
+
+    func userInboundEventTriggered(ctx: ChannelHandlerContext, event: Any) {
+        switch event {
+        case let evt as ChannelEvent where evt == ChannelEvent.inputClosed:
+            self.reads.append(.halfClose)
+        default:
+            ctx.fireUserInboundEventTriggered(event)
+        }
+    }
+}
+
+private final class SuppressingReadsHandler: ChannelOutboundHandler {
```

Comment:
IMHO the name is confusing as it really does not suppress reads but just count these and forward the reads to the next handler. Maybe call it `ReadCountHandler` ?